lib/getMongoCachedJSONFetcher.js

```diff
@@ -45,39 +47,61 @@ const doFetch = async (jsonUrl, options) => {
  * @param cacheLifetime
  * @param fetchOptions
  * @param transform
+ * @param timeout
  * @returns {Promise<{
  *     content: *|undefined,
  *     error: Error|null
  * }>}
  */
-const get = async (jsonUrl, collection, { cacheLifetime, fetchOptions, transform }) => {
+
+const get = async (
+    jsonUrl,
+    collection,
+    {
+        cacheLifetime = 10000,
+        fetchOptions = {},
+        transform = val => val,
+        timeout = 2000
+    }
+) => {
 
     const cachedFile = await collection.findOne({ _id: jsonUrl });
 
     if (cachedFile && cachedFile.fetchedAt >= new Date(Date.now() - cacheLifetime)) {
         return cachedFile.content;
     }
 
-    try {
-        let fetchResult = await doFetch(jsonUrl, fetchOptions);
+    const constructError = err => new NoOrOldCacheError(err, !!cachedFile, cachedFile && cachedFile.content);
 
-        fetchResult = await transform(fetchResult);
+    return new Promise(async (resolve, reject) => {
 
-        await collection.replaceOne(
-            { _id: jsonUrl },
-            {
-                content: fetchResult,
-                fetchedAt: new Date()
-            },
-            { upsert: true }
-        );
+        const timeoutId = setTimeout(() => {
```

Comment:
This will indeed terminate the method, but will not dispose waiting for "doFetch" result. The code after "doFetch" will eventually get called even though result was rejected previously. Reject in this case will not get rid of the context of the method from memory in fact and will not terminate it like "return" statement does for instance.

Reply:
It is can be simulated by using some other promise instead of "doFetch" and placing larger timeout to it before resolving.

Reply:
It can in fact be seen in the test scenario you created yourself, if you put console.log with some unique id to the function call to both setTimeout callback and after the "doFetch" execution, it can be observed that the code gets executed.

I am not sure if this is intentional or not, but from my perspective it is sort of side-effect so i am just trying to point that one out.

Reply:
Yes, this is intentional scenario as covered in the test, see "The cache should be filled with a fresh content once the request finishes in background." assert.
This mechanism is designed for getting results from APIs with long response time and short time we can wait. If we just stop the request with standard timeout, we'd never update the content. And that's the reason you can provide "timeout" on top-level function options and "fetchOptions" at the same time.

👍  for your sharp eyes anyway :)

Reply:
Allright then. :)